File: implement-shell-tools/wc/wc.js
const fs = require("fs");

// Get arguments after "node wc.js"
const args = process.argv.slice(2);

let countLines = false;
let countWords = false;
let countBytes = false;

let files = [];

// Read arguments
for (let arg of args) {
  if (arg === "-l") {
    countLines = true;
  } else if (arg === "-w") {
    countWords = true;
  } else if (arg === "-c") {
    countBytes = true;
  } else {
    files.push(arg);
  }
}

// If no flags are given, wc shows all three
if (!countLines && !countWords && !countBytes) {
  countLines = true;
  countWords = true;
  countBytes = true;
}

// Totals
let totalLines = 0;
let totalWords = 0;
let totalBytes = 0;
let filesCounted = 0;

// Function to count one file
function countFile(fileName) {
  try {
    const content = fs.readFileSync(fileName, "utf8");

    let lines = content.split("\n").length - 1;

    let words = content
      .trim()
      .split(/\s+/)
      .filter((word) => word.length > 0).length;

    let bytes = Buffer.byteLength(content);

    totalLines += lines;
    totalWords += words;
    totalBytes += bytes;
    filesCounted++;

    let result = "";

    if (countLines) {
      result += lines + " ";
    }

    if (countWords) {
      result += words + " ";
    }

    if (countBytes) {
      result += bytes + " ";
    }

    result += fileName;

    console.log(result);
  } catch (error) {
    console.log("Cannot read file: " + fileName);
  }
}

// Run wc for every file
for (let file of files) {
  countFile(file);
}

// Print totals if more than one file was counted
if (filesCounted > 1) {
  let result = "";

  if (countLines) {
    result += totalLines + " ";
  }

  if (countWords) {
    result += totalWords + " ";
  }

  if (countBytes) {
    result += totalBytes + " ";
  }

  result += "total";

  console.log(result);
}

File: implement-shell-tools/wc/wc.py
import sys


# Get everything the user typed after wc.py
arguments = sys.argv[1:]


# These are our options
count_lines = False
count_words = False
count_characters = False


# Store file names here
filenames = []


# Check every argument
for argument in arguments:

    if argument == "-l":
        count_lines = True

    elif argument == "-w":
        count_words = True

    elif argument == "-c":
        count_characters = True

    else:
        filenames.append(argument)



# If the user did not give any flag,
# show all counts like normal wc
if (not count_lines 
        and not count_words 
        and not count_characters):

    count_lines = True
    count_words = True
    count_characters = True



# Go through every file
for filename in filenames:

    # Open the file
    with open(filename, "r") as file:

        # Read the whole file
        content = file.read()


        # Count lines
        lines = len(content.splitlines())


        # Count words
        words = len(content.split())


        # Count characters
        characters = len(content)



        # Prepare the answer
        answer = ""


        if count_lines:
            answer += str(lines) + " "


        if count_words:
            answer += str(words) + " "


        if count_characters:
            answer += str(characters) + " "



        # Print the result
        print(answer + filename)